### PR TITLE
Asynch patient fetching

### PR DESCRIPTION
**Pull request by Dtphelan1**
This  PR adds in the functionality to support asynchronous loading of patient data from a data source. In order to do so, a few main changes had to be made: 
- Main dashboards are only ever loaded when there is a non-empty patient in the local state. This avoids having to make the dashboards sophisticated enough to handle conditionally interacting with and using patient data. 
- The PatientSummaryHeader, on the other hand, _was_ changed to handle having no patient data. In this case, it only renders the logo and the current user information. 
- We can now display loadingAnimations and loadingErrors. These are always rendered, but only fade into the display when loading is happening in the first case and when an error occured in loading patient information in the second. 
- Added 'Gestalt' information to each of our data sources, describing what operations are supported and what kinds of requestTypes are supported.  On the origins of this term see [here](https://en.wikipedia.org/wiki/Gestalt_(Mac_OS). For each dataSource, this defines what kind of requests can be made and whether or not the request is synchronous or asynchronous. At the app level, we use that information to determine whether or not to send a callback along to get the patient data. 
- **Need to `yarn install`** since I updated our version of material ui from `1.0.0-beta.30` to `10.0.-beta.32`. We needed this upgrade to fix a bug in the `Fade` component  that gets used to fade in and out the application on loading. This was the absolute minimum requirement necessary, as evidenced by the changelog found [here](https://github.com/mui-org/material-ui/blob/master/CHANGELOG.md#component-fixes--enhancements-33). 
    - **N.B.** This introduced a new breaking change related to the `Button` component that I have since fixed, described in the list [here](https://github.com/mui-org/material-ui/blob/master/CHANGELOG.md#breaking-change-32).

For Testing: 
- Visit `pilot1` to confirm that synchronous loading of hardcoded data still works (timeout added for effect) 
- Visit `ccp` to confirm that synchronous loading of hardcoded data still works (no timeout)
- Visit `p2` with the local node server up to confirm that synchronous loading of remote data still works. 
- Visit `p2` without the local node server up to confirm that errors are displayed well in the sync case. 
- Visit `p2` with the local node server up, AND modify the local API file in `node_modules/shr_rest_client/src/api/DefaultApi.js` to use the asynchronous fetch to confirm that errors are displayed well in the asynch case. 
- We obviously can't confirm that async fetch itself works flawlessly until we have such a resource

### NOTE: I've left in `setTimeout`'s throughout the loading of patient data so that reviewers can judge the look/feel of the loading animations. These **need to be removed** before merging. 

_Pull requests into Flux require the following. Submitter should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:. Reviewers should complete every item on the bulleted list when reviewing._

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [x] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [x] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [x] Branch implements task as expected
- [x] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- N/A Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- N/A Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- N/A Document any new APIs/extension points
- N/A Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [x] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- N/A Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

File: src/dataaccess/IDataSource.jsx [hunk 1]
```diff
@@ -13,6 +13,23 @@ class IDataSource {
...
         throw new Error("savePatient not implemented by " + this.constructor.name);
     }
+
+    // On the origins of this term see: https://en.wikipedia.org/wiki/Gestalt_(Mac_OS)
+    getGestalt() { 
+        // 
+        // Shape of the gestalt : { 
+        //      requestTypes: {
+        //          sync: bool
+        //          asynch: bool
```
Comment (dehall) [inline src/dataaccess/IDataSource.jsx]: Quick note that the comment here says `asynch` but everywhere else in the code it's `async`.

File: src/dataaccess/DataAccess.jsx [hunk 2]
```diff
@@ -36,4 +36,8 @@ export default class DataAccess {
     savePatient(patient) {
```
Comment (dehall) [inline src/dataaccess/DataAccess.jsx]: I understand that it's not part of this task, but at some point the other functions should probably be updated to support sync/async as well. As a heads-up the SMART on FHIR client library makes all service calls asynchronously, so the savePatient function in that upcoming data source will have to be async.

Comment (Dtphelan1) [reply src/dataaccess/DataAccess.jsx]: Agreed. I can make a JIRA to support thos additional functions. Thanks Dylan!

File: src/dataaccess/IDataSource.jsx (hunk 1 above)
Comment (Dtphelan1) [reply src/dataaccess/IDataSource.jsx]: Taken care of! 👍

File: src/containers/FullApp.jsx [hunk 3]
```diff
@@ -148,13 +157,56 @@ export class FullApp extends Component {
...
+                    patient, 
+                    loading: false,
+                }), 5000);
+                //
+                // TODO: RESTORE AFTER TESTING
+                // this.setState({
+                //     patient, 
+                //     loading: false
+                // });
+            } catch (error) {
```
Comment (dehall) [inline src/containers/FullApp.jsx]: If there is an error somewhere, nothing from here on gives you easy access to the full stack trace, unless you're already debugging. Same with the async path above. The LoadingError object just shows the message which isn't always sufficient.
I'm not sure if there's a Flux standard for error logging, but it could be helpful to write the error to console.error either here or in the LoadingError class, that way all the individual DataSource classes don't necessarily have to wrap everything in try/catch with a logger. Compare to before when there was no error handler and any errors would just bubble up.

Comment (Dtphelan1) [reply src/containers/FullApp.jsx]: I like it -- implementing now.

Comment (Dtphelan1) [reply src/containers/FullApp.jsx]: Nice. I added that functionality into the different apps, so that's logged properly as errors.

File: src/dataaccess/RestApiDataSource.jsx [hunk 4]
```diff
@@ -6,14 +6,32 @@ class RestApiDataSource extends IDataSource {
...
-        // return this.api.getPatientById(id, callback);
-        
+
+    getGestalt() { 
+        return this._gestalt
+    }
+
+    getPatient(id, callback) {
+        // Asynchronous call
+        // return this.api.getPatientById(id, (error, data, response) => {
```
Comment (dehall) [inline src/dataaccess/RestApiDataSource.jsx]: Should this still be commented? In the description of the task it says:
> This will also involve changing the shr-javascript-client library to remove the additional code to make a synchronous call.
> When this is working, the commented out test in DataAccess.test.js for getting a patient from the RestApiDataSource should be uncommented and should pass.

So unless something changed I think this async version should be uncommented and that test case should be uncommented and changed to use the callback method. To correctly test an async call it should look something like this where it calls `done` in the callback:
```
    it('should do a thing async', function(done) {
        dataAccess.getPatient(id, (result, error) => {
            expect(...);
            done();
        });
    });
```

Comment (Dtphelan1) [reply src/dataaccess/RestApiDataSource.jsx]: Great observation. From the discussions I had with @gregquinn2001, we would have liked to have gotten the RestApiDataSource working but we didn't want to hold up this task if we couldn't. In this case, I made changes to the shr_rest_client for testing but ran into a litany of CORS issues that were relentless. Because of that, I decided to punt on that for now, particularly because our next focus should be on getting @dehall's new data source up and running and not this older data source.

tl;dr Tried to fix this async; ran into CORS issues; punting because this dataSource isn't a priority

**Comment by Dtphelan1** [comment]
Bump on @gregquinn2001  👍  Taking accountability on my PR 😄

**Comment by Dtphelan1** [comment]
Tried a new shape to the gestalt -- take a peek! 👍

**Comment by gregquinn2001** [comment]
Looks good. 1 thought though. We don't need isSupported really right? if async and sync or both false then the operation isn't supported at all? Too subtle?

**Comment by gregquinn2001** [comment]
needs a rebase?

**Comment by Dtphelan1** [comment]
Rebasing and removing the isSupported flag 👍

**Comment by Dtphelan1** [comment]
All set on the rebase. Removing the testing timeouts and then I think we should be good to merge
